# ì½”ë“œ í’ˆì§ˆ ê·œì¹™

**ìµœì¢… ì—…ë°ì´íŠ¸**: 2025-10-21
**ì ìš© ëŒ€ìƒ**: ëª¨ë“  Agent ë° ëª¨ë“  Python ì½”ë“œ

---

## 1. íŒŒì¼ ë¼ì¸ ìˆ˜ ì œí•œ

### ê¸°ë³¸ ê·œì¹™:
- **ì‹ ê·œ íŒŒì¼ì€ 1000ì¤„ ì´ë‚´ë¥¼ ëª©í‘œë¡œ ì‘ì„±**
- **ìµœëŒ€ 1500ì¤„ê¹Œì§€ í—ˆìš©** (1500ì¤„ ì´ˆê³¼ ì‹œ ë°˜ë“œì‹œ ëª¨ë“ˆ ë¶„ë¦¬)
- íŒŒì¼ ìƒì„±/ìˆ˜ì • ì‹œ ë¼ì¸ ìˆ˜ ìë™ ì²´í¬
- ê¸°ì¡´ íŒŒì¼ì€ ì ì§„ì ìœ¼ë¡œ 1000ì¤„ ì´í•˜ë¡œ ë¦¬íŒ©í† ë§

### ì˜ˆì™¸ ì‚¬í•­:
- **ë©”ì¸ ì‹¤í–‰ íŒŒì¼** (main_auto_trade.py ë“±): 2000ì¤„ê¹Œì§€ í—ˆìš©
- **í†µí•© ì„¤ì • íŒŒì¼** (CLAUDE.md ë“±): ë¶„ë¦¬ ê¶Œì¥í•˜ì§€ë§Œ ì˜ˆì™¸ ê°€ëŠ¥

### ë¼ì¸ ìˆ˜ ì´ˆê³¼ ì‹œ ì¡°ì¹˜:
1. **1000-1500ì¤„**: ê²½ê³  ë°œìƒ, ë‹¤ìŒ ì‘ì—… ì‹œ ë¶„ë¦¬ ê²€í† 
2. **1500ì¤„ ì´ˆê³¼**: ì¦‰ì‹œ ëª¨ë“ˆ ë¶„ë¦¬ í•„ìˆ˜
   - ê´€ë ¨ ê¸°ëŠ¥ë³„ë¡œ ë³„ë„ íŒŒì¼ ìƒì„±
   - ê³µí†µ ìœ í‹¸ë¦¬í‹°ëŠ” `utils/` í´ë”ë¡œ ë¶„ë¦¬
   - í´ë˜ìŠ¤ë³„ë¡œ íŒŒì¼ ë¶„ë¦¬ ê²€í† 

### ëª¨ë“ˆ ë¶„ë¦¬ ì˜ˆì‹œ:
```python
# Before: monolithic_service.py (2000 lines)
class BacktestService:
    def __init__(self): ...
    def run_backtest(self): ...
    def calculate_metrics(self): ...
    def generate_report(self): ...
    def save_results(self): ...

# After: Split into modules
# backtest_service.py (500 lines) - Main orchestration
# backtest_metrics.py (400 lines) - Metrics calculation
# backtest_reporter.py (400 lines) - Report generation
# backtest_storage.py (300 lines) - Result storage
```

---

## 2. ì´ëª¨ì§€ ì‚¬ìš© ê¸ˆì§€

### í•µì‹¬ ê·œì¹™:
**Python ì½”ë“œ, ë¡œê·¸ ë©”ì‹œì§€, ì£¼ì„ì— ì´ëª¨ì§€(âœ…, âŒ, ğŸš€ ë“±) ì‚¬ìš© ì ˆëŒ€ ê¸ˆì§€**

### ì´ìœ :
- **Windows cp949 ì¸ì½”ë”© í™˜ê²½ì—ì„œ UnicodeEncodeError ë°œìƒ ë°©ì§€**
- í¬ë¡œìŠ¤ í”Œë«í¼ í˜¸í™˜ì„± ë³´ì¥
- ë¡œê·¸ íŒŒì¼ì˜ ê°€ë…ì„± ìœ ì§€

### ëŒ€ì²´ í‘œí˜„:
| ì´ëª¨ì§€ | ëŒ€ì²´ í…ìŠ¤íŠ¸ |
|-------|-----------|
| âœ… | [OK] |
| âŒ | [FAIL] ë˜ëŠ” [ERROR] |
| âš ï¸ | [WARNING] |
| â„¹ï¸ | [INFO] |
| ğŸš€ | [STARTED] |
| ğŸ | [COMPLETED] |
| ğŸ“Š | [DATA] |
| ğŸ’° | [PRICE] |
| ğŸ“ˆ | [UP] |
| ğŸ“‰ | [DOWN] |

### ì˜¬ë°”ë¥¸ ì‚¬ìš© ì˜ˆì‹œ:

```python
# âŒ ì˜ëª»ëœ ì˜ˆ
print("âœ… Backtest completed successfully!")
logger.info("ğŸš€ Starting trading system...")
print(f"ğŸ“Š Total trades: {total_trades}")

# âœ… ì˜¬ë°”ë¥¸ ì˜ˆ
print("[OK] Backtest completed successfully!")
logger.info("[STARTED] Starting trading system...")
print(f"[DATA] Total trades: {total_trades}")
```

### ì˜ˆì™¸:
- **ë¬¸ì„œ íŒŒì¼(.md, .txt)ì—ì„œëŠ” ì´ëª¨ì§€ ì‚¬ìš© ê°€ëŠ¥**
- Markdown ë¬¸ì„œì—ì„œ ì‹œê°ì  êµ¬ë¶„ì„ ìœ„í•œ ì´ëª¨ì§€ëŠ” í—ˆìš©

---

## 3. ëª¨ë“ˆ ì¸í„°í˜ì´ìŠ¤ ê´€ë¦¬

### ê¸°ë³¸ ê·œì¹™:
- **ëª¨ë“  LayerëŠ” ì¸í„°í˜ì´ìŠ¤ ë¬¸ì„œ(MD)ë¥¼ ë°˜ë“œì‹œ ì‘ì„±**
- **ëª¨ë“ˆ ê°„ í†µì‹ ì€ ë¬¸ì„œí™”ëœ ì¸í„°í˜ì´ìŠ¤ë¥¼ í†µí•´ì„œë§Œ ìˆ˜í–‰**
- ê° Layerì˜ ë‹´ë‹¹ Agentê°€ ë¬¸ì„œë¥¼ ê¸°ì–µí•˜ê³  ê´€ë¦¬

### ì¸í„°í˜ì´ìŠ¤ ë¬¸ì„œ ìœ„ì¹˜:
```
docs/interfaces/
â”œâ”€â”€ HELPER_LAYER_INTERFACE.md
â”œâ”€â”€ DATABASE_LAYER_INTERFACE.md
â”œâ”€â”€ INDICATOR_LAYER_INTERFACE.md
â”œâ”€â”€ STRATEGY_LAYER_INTERFACE.md
â”œâ”€â”€ SERVICE_LAYER_INTERFACE.md
â””â”€â”€ REPORT_LAYER_INTERFACE.md
```

### ì¸í„°í˜ì´ìŠ¤ ë³€ê²½ ì‹œ:
1. **Orchestratorì—ê²Œ ë³€ê²½ ìš”ì²­**
2. **ì˜í–¥ ë°›ëŠ” ëª¨ë“  Agentì™€ í˜‘ì˜**
3. **ìŠ¹ì¸ í›„ ë³€ê²½ ì‹¤í–‰**
4. **ì¸í„°í˜ì´ìŠ¤ ë¬¸ì„œ ì¦‰ì‹œ ì—…ë°ì´íŠ¸**
5. **ê´€ë ¨ Agentì—ê²Œ ë³€ê²½ì‚¬í•­ í†µë³´**

---

## 4. ì½”ë”© ìŠ¤íƒ€ì¼

### Python ì½”ë”© í‘œì¤€:
- **PEP 8** ì¤€ìˆ˜
- **Type hints** ì ê·¹ ì‚¬ìš©
- **Docstrings** í•„ìˆ˜ ì‘ì„±

### Docstring í˜•ì‹:
```python
def calculate_signal(
    df_daily: pd.DataFrame,
    df_weekly: pd.DataFrame,
    threshold: float = 0.5
) -> Dict[str, Any]:
    """
    Calculate trading signal based on daily and weekly data.

    Args:
        df_daily: Daily price data with OHLCV columns
        df_weekly: Weekly price data with technical indicators
        threshold: Minimum signal strength threshold (default: 0.5)

    Returns:
        Dictionary containing:
            - final_signal: SignalType enum (BUY/SELL/HOLD)
            - signal_strength: float (0.0 to 1.0)
            - confidence: float (0.0 to 1.0)
            - target_price: float
            - losscut_price: float

    Raises:
        ValueError: If dataframes are empty or missing required columns

    Example:
        >>> signal = calculate_signal(df_D, df_W, threshold=0.6)
        >>> print(signal['final_signal'])
        SignalType.BUY
    """
    # Implementation
    pass
```

### ë³€ìˆ˜ ëª…ëª… ê·œì¹™:
```python
# âœ… ì˜¬ë°”ë¥¸ ì˜ˆ
df_daily = load_daily_data()
signal_strength = calculate_strength()
is_backtest = True
MAX_POSITIONS = 10

# âŒ ì˜ëª»ëœ ì˜ˆ
d = load_daily_data()  # ë„ˆë¬´ ì§§ìŒ
signalStrength = calculate_strength()  # camelCase ëŒ€ì‹  snake_case ì‚¬ìš©
ib = True  # ì˜ë¯¸ ë¶ˆëª…í™•
maxPositions = 10  # ìƒìˆ˜ëŠ” UPPER_CASE
```

---

## 5. ì—ëŸ¬ ì²˜ë¦¬

### ê¸°ë³¸ ì›ì¹™:
- **ëª¨ë“  ì™¸ë¶€ API í˜¸ì¶œì€ try-exceptë¡œ ê°ì‹¸ê¸°**
- **êµ¬ì²´ì ì¸ ì˜ˆì™¸ íƒ€ì… ì§€ì •**
- **ì—ëŸ¬ ë¡œê·¸ ë‚¨ê¸°ê¸°**
- **ì ì ˆí•œ ê¸°ë³¸ê°’ ë˜ëŠ” fallback ì œê³µ**

### ì—ëŸ¬ ì²˜ë¦¬ ì˜ˆì‹œ:
```python
# âœ… ì˜¬ë°”ë¥¸ ì˜ˆ
def load_market_data(symbol: str) -> pd.DataFrame:
    """Load market data with proper error handling"""
    try:
        data = fetch_from_api(symbol)
        if data is None or data.empty:
            logger.warning(f"[WARNING] No data for {symbol}, returning empty DataFrame")
            return pd.DataFrame()
        return data

    except ConnectionError as e:
        logger.error(f"[ERROR] Connection failed for {symbol}: {e}")
        return pd.DataFrame()

    except ValueError as e:
        logger.error(f"[ERROR] Invalid data for {symbol}: {e}")
        return pd.DataFrame()

    except Exception as e:
        logger.error(f"[ERROR] Unexpected error for {symbol}: {e}")
        return pd.DataFrame()

# âŒ ì˜ëª»ëœ ì˜ˆ
def load_market_data(symbol: str) -> pd.DataFrame:
    """No error handling - dangerous!"""
    data = fetch_from_api(symbol)  # What if this fails?
    return data
```

---

## 6. ë¡œê¹… ê·œì¹™

### ë¡œê·¸ ë ˆë²¨:
- **DEBUG**: ê°œë°œ ì¤‘ ë””ë²„ê¹… ì •ë³´
- **INFO**: ì¼ë°˜ ì‹¤í–‰ íë¦„ ì •ë³´
- **WARNING**: ê²½ê³  (ì²˜ë¦¬ëŠ” ê³„ì†ë¨)
- **ERROR**: ì—ëŸ¬ (ì¼ë¶€ ê¸°ëŠ¥ ì‹¤íŒ¨)
- **CRITICAL**: ì¹˜ëª…ì  ì˜¤ë¥˜ (ì‹œìŠ¤í…œ ì¤‘ë‹¨)

### ë¡œê¹… ì˜ˆì‹œ:
```python
import logging
logger = logging.getLogger(__name__)

# DEBUG
logger.debug("[DEBUG] DataFrame shape: %s", df.shape)

# INFO
logger.info("[INFO] Loaded %d symbols from database", len(symbols))

# WARNING
logger.warning("[WARNING] Missing data for %s, using fallback", symbol)

# ERROR
logger.error("[ERROR] Failed to calculate signal for %s: %s", symbol, e)

# CRITICAL
logger.critical("[CRITICAL] MongoDB connection lost, system halted")
```

### ë¡œê·¸ ë©”ì‹œì§€ í˜•ì‹:
```python
# âœ… ì˜¬ë°”ë¥¸ ì˜ˆ
logger.info("[INFO] Backtest completed: %.2f%% return", total_return * 100)
logger.error("[ERROR] Symbol %s failed validation: %s", symbol, reason)

# âŒ ì˜ëª»ëœ ì˜ˆ
logger.info(f"Backtest completed: {total_return * 100}%")  # f-string ëŒ€ì‹  %s ì‚¬ìš©
logger.error("Error!")  # êµ¬ì²´ì  ì •ë³´ ì—†ìŒ
```

---

## 7. í…ŒìŠ¤íŠ¸ ì½”ë“œ

### í…ŒìŠ¤íŠ¸ íŒŒì¼ ìœ„ì¹˜:
- ëª¨ë“  `test_*.py` íŒŒì¼ì€ `Test/` í´ë”ì— ë°°ì¹˜
- í†µí•© í…ŒìŠ¤íŠ¸: `Test/integration/`
- ìœ ë‹› í…ŒìŠ¤íŠ¸: `Test/unit/`

### í…ŒìŠ¤íŠ¸ ì»¤ë²„ë¦¬ì§€:
- **í•µì‹¬ ë¡œì§**: 80% ì´ìƒ
- **ìœ í‹¸ë¦¬í‹° í•¨ìˆ˜**: 60% ì´ìƒ
- **UI/CLI ì½”ë“œ**: 30% ì´ìƒ

### í…ŒìŠ¤íŠ¸ ëª…ëª… ê·œì¹™:
```python
def test_signal_generation_with_valid_data():
    """Test that signal generation works with valid input"""
    pass

def test_signal_generation_with_empty_data():
    """Test that signal generation handles empty dataframe"""
    pass

def test_signal_generation_with_missing_columns():
    """Test that signal generation raises error for missing columns"""
    pass
```

---

## 8. Git Commit ë©”ì‹œì§€

### Commit ë©”ì‹œì§€ í˜•ì‹:
```
<type>: <subject>

<body>

<footer>
```

### Type:
- **feat**: ìƒˆë¡œìš´ ê¸°ëŠ¥ ì¶”ê°€
- **fix**: ë²„ê·¸ ìˆ˜ì •
- **docs**: ë¬¸ì„œ ìˆ˜ì •
- **refactor**: ì½”ë“œ ë¦¬íŒ©í† ë§
- **test**: í…ŒìŠ¤íŠ¸ ì¶”ê°€/ìˆ˜ì •
- **chore**: ë¹Œë“œ/ì„¤ì • ë³€ê²½

### ì˜ˆì‹œ:
```
feat: Add real-time monitoring for held positions

- Monitor buy signal stocks (top 10)
- Monitor currently held positions
- Display separate lists for each category

Agents: Service
Ref: main_auto_trade.py:2087-2131
```

---

## 9. ì½”ë“œ ë¦¬ë·° ì²´í¬ë¦¬ìŠ¤íŠ¸

### ì œì¶œ ì „ í™•ì¸ì‚¬í•­:
- [ ] íŒŒì¼ ë¼ì¸ ìˆ˜ 1500ì¤„ ì´í•˜
- [ ] ì´ëª¨ì§€ ì‚¬ìš©í•˜ì§€ ì•ŠìŒ
- [ ] ëª¨ë“  í•¨ìˆ˜ì— docstring ì‘ì„±
- [ ] Type hints ì¶”ê°€
- [ ] ì—ëŸ¬ ì²˜ë¦¬ êµ¬í˜„
- [ ] ë¡œê·¸ ë©”ì‹œì§€ í¬í•¨
- [ ] í…ŒìŠ¤íŠ¸ ì½”ë“œ ì‘ì„±
- [ ] ì¸í„°í˜ì´ìŠ¤ ë¬¸ì„œ ì—…ë°ì´íŠ¸

---

## ì°¸ì¡° ë¬¸ì„œ

- **Agent í˜‘ì—…**: `docs/rules/AGENT_COLLABORATION.md`
- **íŒŒì¼ ê¶Œí•œ**: `docs/rules/FILE_PERMISSIONS.md`
- **MongoDB ê·œì¹™**: `docs/rules/MONGODB_RULES.md`
